## First contact

### Imports

In [1]:
import pandas as pd
import plotly.express as px
import plotly.graph_objects as go

### Creating dataframes

In [2]:
df = pd.read_json("datasets/train_dataset.json")
df_test = pd.read_json("datasets/test_dataset.json")
df_val = pd.read_json("datasets/val_dataset.json")

### First look inside the data

##### **Dataset Columns Description**
1. **`graphic card type`**  
   The type of graphics card in the laptop (e.g., integrated, NVIDIA, AMD). Influences the price and performance in tasks like gaming or graphic editing.

2. **`communications`**  
   Communication technologies available in the laptop (e.g., Wi-Fi, Bluetooth, Ethernet).

3. **`resolution (px)`**  
   Screen resolution of the laptop, specified in pixels (e.g., 1920x1080). Relates to display quality.

4. **`CPU cores`**  
   The number of cores in the processor. Affects performance in multitasking and intensive workloads.

5. **`RAM size`**  
   The amount of RAM in the laptop (e.g., 8GB, 16GB), a crucial factor for overall performance.

6. **`operating system`**  
   The operating system installed on the laptop (e.g., Windows 10, macOS, Linux). May affect the price depending on the licensing.

7. **`drive type`**  
   The type of storage in the laptop (e.g., HDD, SSD). SSDs offer faster performance and are usually more expensive.

8. **`input devices`**  
   Input devices included in the laptop (e.g., backlit keyboard, touchpad, fingerprint reader).

9. **`multimedia`**  
   Multimedia features available in the laptop (e.g., speakers, microphones, webcam).

10. **`RAM type`**  
    The type of RAM installed (e.g., DDR3, DDR4), impacting performance and compatibility.

11. **`CPU clock speed (GHz)`**  
    The clock speed of the processor, measured in GHz. Higher speeds generally indicate better performance for single-threaded tasks.

12. **`CPU model`**  
    The specific model of the processor (e.g., Intel Core i5-1135G7, AMD Ryzen 7 5800H). Provides detailed processor capabilities.

13. **`state`**  
    The condition of the laptop (e.g., new, refurbished, used). Directly impacts the price.

14. **`drive memory size (GB)`**  
    The storage capacity in GB (e.g., 256GB, 1TB). Larger capacities often come at a higher price.

15. **`warranty`**  
    The warranty period provided by the manufacturer (e.g., 1 year, 2 years).

16. **`screen size`**  
    The size of the screen, typically measured in inches (e.g., 14", 15.6"). Larger or higher-quality screens may increase the cost.

17. **`buynow_price`**  
    The buy-now price of the laptop. This is the target variable for the prediction model.

In [3]:
df.head()

,graphic card type,communications,resolution (px),CPU cores,RAM size,operating system,drive type,input devices,multimedia,RAM type,CPU clock speed (GHz),CPU model,state,drive memory size (GB),warranty,screen size,buynow_price
7233,dedicated graphics,"[bluetooth, lan 10/100/1000 mbps]",1920 x 1080,4,32 gb,[no system],ssd + hdd,"[keyboard, touchpad, illuminated keyboard, num...","[SD card reader, camera, speakers, microphone]",ddr4,2.6,intel core i7,new,1250.0,producer warranty,"17"" - 17.9""",4999.0
5845,dedicated graphics,"[wi-fi, bluetooth, lan 10/100 mbps]",1366 x 768,4,8 gb,[windows 10 home],ssd,"[keyboard, touchpad, numeric keyboard]","[SD card reader, camera, speakers, microphone]",ddr3,2.4,intel core i7,new,256.0,seller warranty,"15"" - 15.9""",2649.0
10303,None,"[bluetooth, nfc (near field communication)]",1920 x 1080,2,8 gb,[windows 10 home],hdd,None,[SD card reader],ddr4,1.6,intel core i7,new,1000.0,producer warranty,"15"" - 15.9""",3399.0
10423,None,None,None,2,None,None,None,None,None,None,NaN,None,new,NaN,producer warranty,None,1599.0
5897,integrated graphics,"[wi-fi, bluetooth]",2560 x 1440,4,8 gb,[windows 10 home],ssd,"[keyboard, touchpad, illuminated keyboard]","[SD card reader, camera, speakers, microphone]",ddr4,1.2,other CPU,new,256.0,producer warranty,"12"" - 12.9""",4499.0


In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 4711 entries, 7233 to 6037
Data columns (total 17 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   graphic card type       4417 non-null   object 
 1   communications          4261 non-null   object 
 2   resolution (px)         4361 non-null   object 
 3   CPU cores               4711 non-null   object 
 4   RAM size                4457 non-null   object 
 5   operating system        4335 non-null   object 
 6   drive type              4454 non-null   object 
 7   input devices           4321 non-null   object 
 8   multimedia              4310 non-null   object 
 9   RAM type                4212 non-null   object 
 10  CPU clock speed (GHz)   4181 non-null   float64
 11  CPU model               4389 non-null   object 
 12  state                   4711 non-null   object 
 13  drive memory size (GB)  4439 non-null   float64
 14  warranty                4711 non-null   ob

## EDA

Checking the null values

As we can see, from the 4711 entries, we have some null values in some columns

In [19]:
null_summary = pd.DataFrame({
  "Column": df.columns,
  "Null count": df.isnull().sum(),
  "Percentage Null": ((df.isnull().sum() / len(df)) * 100).round(2)
}).sort_values(by="Null count", ascending=False)

fig = px.bar(null_summary, x="Column", y="Null count",
             title="Null Values by Column", text="Percentage Null",template="plotly_dark")
fig.update_layout(xaxis_title="Columns", yaxis_title="Count of Nulls")
fig.show()

In [ ]:
fig = px.bar_polar(null_summary,
                   r="Null count",
                   theta="Column",
                   color="Null count",
                   title="Null Values by Column (Polar Plot)",
                   template="plotly_dark")

fig.update_traces(hovertemplate="Column: %{theta}<br>Null Count: %{r}<br>Percentage: %{customdata:.2f}%")

fig.update_traces(customdata=null_summary["Percentage Null"])

fig.update_layout(
    polar=dict(
        radialaxis=dict(visible=False),
        angularaxis=dict(showticklabels=True)
    ),
    showlegend=False
)

fig.show()

Price distribution

In [20]:
fig = px.histogram(df, x="buynow_price", nbins=40, title="Distribution of buy-now price", template="plotly_dark")
fig.update_layout(xaxis_title="Price", yaxis_title="Frequency")
fig.update()

Trying to identify outliers

In [21]:
fig = px.box(df, x="buynow_price", title="outliers buy-now price", template="plotly_dark")
fig.update_layout(yaxis_title="Price")
fig.show()